src/audio/audio_sink.cpp

```diff
@@ -11,71 +11,146 @@
 #include <cassert>
 #include <climits>
 #include <cstring>
+#include <memory>
 #include <string>
 
-#include "portaudiocpp/PortAudioCpp.hxx"
+#include "SDL.h"
 
 #include "../errors.hpp"
-#include "../sample_formats.hpp"
 #include "../messages.h"
 #include "audio_sink.hpp"
 #include "audio_source.hpp"
 #include "ringbuffer.hpp"
+#include "sample_formats.hpp"
 
-const size_t AudioSink::RINGBUF_POWER = 16;
+const size_t SdlAudioSink::RINGBUF_POWER = 16;
 
-AudioSink::AudioSink(const AudioSource &source,
-                     const AudioSinkConfigurator &conf)
+/**
+ * The callback used by SDL_Audio.
+ * Trampolines back into vsink, which must point to an SdlAudioSink.
+ */
+static void SDLCallback(void *vsink, std::uint8_t *data, int len)
+{
+	assert(vsink != nullptr);
+	auto sink = static_cast<SdlAudioSink *>(vsink);
+	sink->Callback(data, len);
+}
+
+/* static */ std::unique_ptr<AudioSink> SdlAudioSink::Build(
+                const AudioSource &source, int device_id)
+{
+	return std::unique_ptr<AudioSink>(new SdlAudioSink(source, device_id));
+}
+
+SdlAudioSink::SdlAudioSink(const AudioSource &source, int device_id)
     : bytes_per_sample(source.BytesPerSample()),
       ring_buf(RINGBUF_POWER, source.BytesPerSample()),
       position_sample_count(0),
       just_started(false),
-      input_ready(false)
+      source_out(false),
+      state(Audio::State::STOPPED)
+{
+	const char *name = SDL_GetAudioDeviceName(device_id, 0);
+	if (name == nullptr) {
+		throw ConfigError(std::string("invalid device id: ") +
+		                  std::to_string(device_id));
+	}
+
+	SDL_AudioSpec want;
+	SDL_zero(want);
+	want.freq = source.SampleRate();
+	want.format = SDLFormat(source.OutputSampleFormat());
+	want.channels = source.ChannelCount();
+	want.callback = &SDLCallback;
+	want.userdata = (void *)this;
+
+	SDL_AudioSpec have;
+	SDL_zero(have);
+
+	this->device = SDL_OpenAudioDevice(name, 0, &want, &have, 0);
+	if (this->device == 0) {
+		throw ConfigError(std::string("couldn't open device: ") +
+		                  SDL_GetError());
+	}
+}
+
+SdlAudioSink::~SdlAudioSink()
+{
+	if (this->device == 0) return;
+	SDL_CloseAudioDevice(this->device);
+
+	SdlAudioSink::CleanupLibrary();
+}
+
+/* static */ void SdlAudioSink::InitLibrary()
 {
-	this->stream = decltype(this->stream)(conf.Configure(source, *this));
+	if (SDL_Init(SDL_INIT_AUDIO) != 0) {
+		throw ConfigError(std::string("could not initialise SDL: ") +
+		                  SDL_GetError());
+	}
 }
 
-void AudioSink::Start()
+/* static */ void SdlAudioSink::CleanupLibrary()
```

Comment:
Would then need to keep track of how many times a SdlAudioSink has been constructed and destructed (ew), because the library would need to be cleaned up if, and only if, every single SdlAudioSink ever has disappeared (else SdlAudioSinks'll stop working).  This, or the SdlAudioSink needs to be a singleton (ew).  This, or we have a nasty, but rare, bug.